File: twitbase/src/main/scala/hbase/HBaseUtils.scala
package hbase

import java.security.MessageDigest

import hbase.client.HBaseClient.{getTable, using}
import org.apache.hadoop.hbase.client.{Result, ResultScanner, Scan, Table}
import org.apache.hadoop.hbase.util.Bytes

import scala.collection.JavaConverters._

object HBaseUtils {
  def md5(input: String): Array[Byte] = {
    val e: MessageDigest = MessageDigest.getInstance("MD5")
    e.update(Bytes.toBytes(input))
    e.digest()
  }

  def scan[T](scan: Scan, table: Array[Byte], fn: Result => T): List[T] =
    using[List[T], Table](getTable(table)){t =>
       using[List[T], ResultScanner](t.getScanner(scan)) {scanner =>
         scanner.asScala.map(fn).toList
       }
    }
}

File: twitbase/src/main/java/hbase/HBaseUtils.java
package hbase;

import org.apache.hadoop.hbase.HConstants;
import org.apache.hadoop.hbase.util.Bytes;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

public class HBaseUtils {
    public static byte[] md5(String input ) {
        MessageDigest e;
        try {
            e = MessageDigest.getInstance("MD5");
            e.update(Bytes.toBytes(input));
            return e.digest();
        } catch (NoSuchAlgorithmException e1) {
            throw new RuntimeException(e1);
        }
    }

    public static byte[] nextClosest(byte[] rowKeyPrefix){
        int offset = rowKeyPrefix.length;
        while (offset > 0) {
            if (rowKeyPrefix[offset - 1] != (byte) 0xFF) {
                break;
            }
            offset--;
        }
        if (offset == 0) {
            return HConstants.EMPTY_END_ROW;
        }
        byte[] newStopRow = Arrays.copyOfRange(rowKeyPrefix, 0, offset);
        newStopRow[newStopRow.length - 1]++;
        return newStopRow;
    }
}
